### PR TITLE
Configuration per instance

### PR DESCRIPTION
**Pull request by leonardoalifraco**
Hi @kernow 

I made the changes required for the configuration not to be necessarily singleton. (https://github.com/kernow/shortcode/issues/43)
The singleton interface remains the same, so this PR is not introducing breaking changes.

However, you will be able to instantiate a new processor with a configuration object differently from the one available by singleton.

I'll write a few comments over the changes to explain them better.

### CODE REVIEW

File: lib/shortcode.rb [hunk 1]
```diff
@@ -25,7 +25,7 @@ def self.setup
 
   def self.register_presenter(*presenters)
     presenters.each do |presenter|
-      Shortcode::Presenter.register presenter
+      self.configuration.register_presenter presenter
```
Comment (leonardoalifraco) [inline lib/shortcode.rb]: The presenters are now part of the configuration object.

File: lib/shortcode/configuration.rb [hunk 2]
```diff
@@ -38,5 +41,22 @@ def initialize
...
+
+  private
+
+    def validate_presenter(presenter)
+      raise ArgumentError, "The presenter must define the class method #for" unless presenter.respond_to?(:for)
+      raise ArgumentError, "The presenter must define an initialize method" unless presenter.private_instance_methods(false).include?(:initialize)
+      %w(content attributes).each do |method|
+        raise ArgumentError, "The presenter must define the method ##{method}" unless presenter.method_defined?(method.to_sym)
+      end
+    end
```
Comment (leonardoalifraco) [inline lib/shortcode/configuration.rb]: I have moved the class methods from the Shortcode::Presenter class as instance methods of the Shortcode::Configuration class.

File: lib/shortcode/configuration.rb [hunk 3]
```diff
@@ -38,5 +41,22 @@ def initialize
     @self_closing_tags    = []
     @attribute_quote_type = '"'
     @use_attribute_quotes = true
+    @presenters = {}
```
Comment (leonardoalifraco) [inline lib/shortcode/configuration.rb]: The presenters have an empty hash as default value

File: lib/shortcode/parser.rb [hunk 4]
```diff
@@ -1,36 +1,38 @@
...
-  rule(:block)  { (open >> (block | text | open_close).repeat.as(:inner) >> close) }
-
-  rule(:body) { (block | text | open_close).repeat.as(:body) }
-  root(:body)
+    self.class.rule(:text)   { ((close | block | open_close).absent? >> any).repeat(1).as(:text) }
+    self.class.rule(:block)  { (open >> (block | text | open_close).repeat.as(:inner) >> close) }
 
+    self.class.rule(:body) { (block | text | open_close).repeat.as(:body) }
+    self.class.root(:body)
+  end
```
Comment (leonardoalifraco) [inline lib/shortcode/parser.rb]: All the rules for the parser are evaluated in the constructor, based on the injected configuration.

File: lib/shortcode/processor.rb [hunk 5]
```diff
@@ -1,17 +1,21 @@
...
+
   def process(string, additional_attributes=nil)
     transformer.apply parser.parse(string), additional_attributes: additional_attributes
   end
 
   private
 
     def parser
-      @parser ||= Shortcode::Parser.new
+      @parser ||= Shortcode::Parser.new(@configuration)
```
Comment (leonardoalifraco) [inline lib/shortcode/processor.rb]: The parser instance gets the configuration injected.

File: lib/shortcode/processor.rb [hunk 6]
```diff
@@ -1,17 +1,21 @@
...
     end
 
     def transformer
-      @transformer ||= Shortcode::Transformer.new
+      @transformer ||= Shortcode::Transformer.new(@configuration)
```
Comment (leonardoalifraco) [inline lib/shortcode/processor.rb]: The transformer instance gets the configuration injected.

File: spec/parser_spec.rb [hunk 7]
```diff
@@ -4,7 +4,19 @@
...
+    Shortcode::Configuration.new.tap do |config|
+      config.template_parser = :erb
+      config.template_path = File.join File.dirname(__FILE__), "support/templates/erb"
+      config.templates = nil
+      config.block_tags = [:quote, :collapsible_list, :item, :timeline_person, :rails_helper, :custom_helper]
+      config.self_closing_tags = [:timeline_event, :timeline_info]
+      config.attribute_quote_type = '"'
+      config.use_attribute_quotes = true
+    end
+  }
```
Comment (leonardoalifraco) [inline spec/parser_spec.rb]: Here I build the configuration to be injected in the parser

File: spec/performance_spec.rb [hunk 8]
```diff
@@ -8,34 +8,45 @@
...
+#     Shortcode::Configuration.new.tap do |config|
+#       config.template_parser = :erb
+#       config.template_path = File.join File.dirname(__FILE__), "support/templates/erb"
+#       config.templates = nil
+#       config.block_tags = [:quote, :collapsible_list, :item, :timeline_person, :rails_helper, :custom_helper]
+#       config.self_closing_tags = [:timeline_event, :timeline_info]
+#       config.attribute_quote_type = '"'
+#       config.use_attribute_quotes = true
+#     end
+#   }
```
Comment (leonardoalifraco) [inline spec/performance_spec.rb]: I have also updated the performance specs.

File: spec/spec_helper.rb [hunk 9]
```diff
@@ -17,8 +17,7 @@
 RSpec.configure do |config|
   config.order = "random"
 
-  config.before(:each) do
-    Shortcode::Presenter.presenters = {}
+  config.before(:each, :type => :singleton_setup) do
```
Comment (leonardoalifraco) [inline spec/spec_helper.rb]: I've created a type of examples to load the singleton setup into.

**Comment by kernow** [comment]
@leonardoalifraco thanks for the PR, I've been pretty busy this week so haven't managed to get around to reviewing it yet. I will get onto it as soon as I can. Just wanted to let you know that I will get to it :)

**Comment by leonardoalifraco** [comment]
Thanks @kernow 
Don't hesitate to let me know if you need something on my end. I've used this on a production project and it worked as expected.

**Comment by toddnestor** [comment]
This would be a nice feature to have, any chance of getting this finished off and merged in?

I've started using this gem recently and like it other than the fact that the configuration is universal and I can't have separate Shortcode parsers for different purposes.

**Pull request closed without merging**